concept-art/has-field.sol

```diff
@@ -0,0 +1,52 @@
+data Unit = Unit
+data Pair(a, b) = Pair(a,b)
+
+type uint = word
+type string = word
+type bool = word
+
+data Memory(t) = Memory(Word)
+
+// this lets us link a given field in a struct to its position in it's
+// underlying generic representation as a tuple. 
+class self:Field(prevTypes, ty) {}
+
+// this struct should desugar into the following
+//struct S {
+//  f1 : uint;
+//  f2 : string;
+//  f3 : bool;
+//}
+
+// a type abstraction over tuples
+type s = Pair(uint, Pair(string, bool))
+
+// unique types identifying each field
+type sf1 = Unit
+type sf2 = Unit
+type sf3 = Unit
+
+// Field instances linking each field to it's position in the underlying tuple
+instance Pair(s, sf1):Field(Unit, uint) {}
+instance Pair(s, sf2):Field(Pair(uint, Unit), string) {}
+instance Pair(s, sf3):Field(Pair(uint, Pair(string, Unit)), bool) {}
```

Comment:
In the second two we don't really need the additional Unit:
```suggestion
instance Pair(s, sf2):Field(uint, string) {}
instance Pair(s, sf3):Field(Pair(uint, string), bool) {}
```

Also not sure if for the first one, we won't need a special
```
type FirstField = Unit;
instance Pair(s, sf1):Field(FirstField, uint) {}
```
Since I'm not sure if e.g. for ``Unit``s in storage we should actually assign a unique slot or the like (maybe not, but maybe better to have the option)